### ISSUE TITLE
Performance issues 

### ISSUE THREAD

**Issue by xldenis** [ocaml/setup-ocaml#703]
Thanks for providing this action which gives provides a smooth experience for getting OCaml CI running.

However, I've noticed that his action takes between 2/3-3/4 of my CI time, making my 500 line ocaml project take longer in CI than a 30kloc Rust one! This is rather disappointing because OCaml's compilation speed is one of its strong points but here we really suffer when compared to JS, Rust or other mainstream languages. 

Taking a [recent CI run](https://github.com/xldenis/whycode/actions/runs/6398413747/job/17368504788), total time was 8m41s, which was broken down as follows:
* 3m44s setting up OCaml
* 2m25 running `opam install . --deps-only --locked`
* 2s building the project
* 2m26 tearing down OCaml and caching dune and opam

I don't understand where all this time is going. Additionally, since the action already caches some parts of `opam` I'm hesitant to add my own layer of caching to speed up the dependency installation.

Additionally, the Mac runners suffer a massive penalty in the action taking over 9 minutes just to set up OCaml. I know they are perhaps more resource constrained than the linux runners, but that seems excessive. 

In comparison my Rust project takes a total of 1m46 to install a Rust toolchain and build my project (with cached dependencies), it seems wrong that _Rust_ would be faster than OCaml in CI, what am I doing wrong?

**Comment by arbipher** [comment]
I have a similar concern.

I am new to GitHub CI and I wonder if it's possible to start with an Ubuntu image in which the stable `opam` and `ocaml` and `dune` are already installed, then at least we may skip the _3m44s_ setting time.

**Comment by smorimoto** [comment]
The rough answers are below:

1. The set-up time should be improved in the next opam minor release. You can use pre-release opam via the new input, and it should probably take less than a minute: https://github.com/ocaml/setup-ocaml/pull/704
2. You might be able to speed it up using the dune cache, but at this point there is no way to make it incredibly fast. If you cache the entire switch, it's harder to catch the project corruption due to transitive dependencies, so the action uses the fastest caching strategy while avoiding that.
3. No complaints here!
4. The cleanup, also known as the post process, took a few minutes and was not so much an issue with the action, but an upstream issue with GitHub Actions, and it was fixed in the last PR and is included in the latest release. If you are interested, please check the PR and the issues linked there for details: https://github.com/ocaml/setup-ocaml/pull/724

**Comment by smorimoto** [comment]
In my opinion, the performance issues here can be solved largely by Relocatable Compiler. Conversely, it's the only realistic and safe option we can take. As you can see in this document, the action is already very optimised: https://github.com/ocaml/setup-ocaml/blob/master/packages/setup-ocaml/README.md

**Comment by leostera** [comment]
I recently experienced this the same issue, where running things on a docker image that has a cached layer for OCaml/opam makes things considerably faster.

> @smorimoto:  If you cache the entire switch, it's harder to catch the project corruption due to transitive dependencies, so the action uses the fastest caching strategy while avoiding that.

would it be possible to aggressively cache if an opam lock is present?

**Comment by smorimoto** [comment]
@leostera 

> I recently experienced this the same issue, where running things on a docker image that has a cached layer for OCaml/opam makes things considerably faster.

Yeah, I believe Docker is another good option!

> would it be possible to aggressively cache if an opam lock is present?

It might make things faster if you could cache the whole thing with the lock file as the cache key, but I'm also afraid that this could cause unintended issues. Unless you use Nix or something like that, I feel that breaking the switch due to a change in the base image will be something you have to consider 🤷‍♂️

**Comment by leostera** [comment]
Right, I'm looking for an 80/20 solution here that'd give me the best performance if:
* the image is fixed
* the ocaml version is fixed
* every package is in the lock file

I'd be okay with this cache breaking if I say bump from ubuntu 20 to 23, or from ocaml 4.14 to 5.1.

**Comment by smorimoto** [comment]
You can use any solution for that purpose. However, the action only caches the bare minimum because I believe it's important to keep the cache size to the bare minimum. This is effective as the number of matrices grows.

**Comment by smorimoto** [comment]
In the last few releases, the speed should be up by tens of seconds, particularly if you use pre-release opam: https://github.com/ocaml/setup-ocaml/releases

**Comment by smorimoto** [comment]
In the latest release, using `allow-prerelease-opam`, you can set up your environment in around 30 seconds.

**Comment by leostera** [comment]
Damn 🤩 great work @smorimoto! 👏🏼 I'll give it a shot.

**Comment by smorimoto** [comment]
Note that using something like dune cache requires you to install dune, which at first glance seems like a lot of set up time!
Also, due to the instability of the GHA cache server, it sometimes takes about a minute 😇 But in my experience, this is less frequent than I thought.

**Comment by smorimoto** [comment]
It's still not incredibly fast on macOS, but it's just faster in the latest release, by reducing unnecessary invocation of brew.

| Before | After |
| :-: | :-: |
| <img width="423" alt="Screenshot 2023-12-09 at 23 52 53" src="https://github.com/ocaml/setup-ocaml/assets/38746192/0b798b04-b93a-49e8-9edc-356b8fcc1fd3"> | <img width="423" alt="Screenshot 2023-12-09 at 23 53 32" src="https://github.com/ocaml/setup-ocaml/assets/38746192/cbc7af82-d97d-4e40-90ac-0d02986f6392"> |
